lib/segment/src/common/reciprocal_rank_fusion.rs

```diff
@@ -0,0 +1,168 @@
+use std::collections::hash_map::Entry;
+
+use ahash::{HashMap, HashMapExt};
+use ordered_float::OrderedFloat;
+
+use crate::types::{ExtendedPointId, ScoredPoint};
+
+/// Reciprocal Rank Fusion (RRF) is a method for combining rankings from multiple sources.
+/// See https://plg.uwaterloo.ca/~gvcormac/cormacksigir09-rrf.pdf
+
+/// Mitigates the impact of high rankings by outlier systems
+const RFF_RANKING_K: f32 = 2.0;
+
+/// Compute RRF scores for multiple results from different sources.
+/// Each response can have a different length.
+/// Each response must be sorted by score in descending order.
+///
+/// The output is a single sorted list of ScoredPoint.
+/// Does not break ties.
+#[allow(dead_code)]
+pub fn rrf_scoring(responses: Vec<Vec<ScoredPoint>>, limit: usize) -> Vec<ScoredPoint> {
+    // ensure each input is sorted
+    debug_assert!(responses
+        .iter()
+        .all(|r| r.windows(2).all(|w| w[0].score >= w[1].score)));
+    fn position_score(position: usize) -> f32 {
+        1.0 / (position as f32 + RFF_RANKING_K)
+    }
+
+    // track RFF scores per point id
+    let mut scores_by_id: HashMap<ExtendedPointId, f32> = HashMap::new();
+
+    // track scored points by id
+    let mut points_by_id: HashMap<ExtendedPointId, ScoredPoint> = HashMap::new();
+
+    for response in responses {
+        for (pos, score) in response.into_iter().enumerate() {
+            let rrf_score = position_score(pos);
+            match scores_by_id.entry(score.id) {
+                Entry::Occupied(mut entry) => {
+                    // accumulate score
+                    *entry.get_mut() += rrf_score;
+                }
+                Entry::Vacant(entry) => {
+                    // init score
+                    entry.insert(rrf_score);
+                    // track point
+                    points_by_id.insert(score.id, score);
+                }
+            }
+        }
+    }
+
+    // sort by tracked scores
+    let mut scores = scores_by_id.into_iter().collect::<Vec<_>>();
+    scores.sort_by(|a, b| OrderedFloat(b.1).cmp(&OrderedFloat(a.1)));
```

Comment:
Might be a little better to allocate the sorted vec instead.
```suggestion
    let mut scores = scores_by_id
        .into_iter()
        .sorted_by(|a, b| OrderedFloat(b.1).cmp(&OrderedFloat(a.1)))
        .collect::<Vec<_>>();
```

**Additional note**
For congruence tests, does it matter if we choose unstable sorting? With RRF we expect same-scored points to occur often

Reply:
I agree that there will be a lot of same scored points.

Going with unstable will make it faster but will make the congruence tests less easy.

I think we should go for unstable.